### PR TITLE
8284563: AArch64: bitperm feature detection for SVE2 on Linux

### PR DESCRIPTION
**Pull request by theRealELiu**
This patch adds BITPERM feature detection for SVE2 on Linux. BITPERM is
an optional feature in SVE2 [1]. It enables 3 SVE2 instructions (BEXT,
BDEP, BGRP). BEXT and BDEP map efficiently to some vector operations,
e.g., the compress and expand functionalities [2] which are proposed in
VectorAPI's 4th incubation [3]. Besides, to generate specific code based
on different architecture features like x86, this patch exports
VM_Version::supports_XXX() for all CPU features.  E.g.,
VM_Version::supports_svebitperm() for easy use.

This patch also fixes a trivial bug, that sets UseSVE back to 1 if it's
2 in SVE1 system.

[1] https://developer.arm.com/documentation/ddi0601/2022-03/AArch64-Registers/ID-AA64ZFR0-EL1--SVE-Feature-ID-register-0
[2] https://bugs.openjdk.java.net/browse/JDK-8283893
[3] https://bugs.openjdk.java.net/browse/JDK-8280173

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284563](https://bugs.openjdk.java.net/browse/JDK-8284563): AArch64: bitperm feature detection for SVE2 on Linux

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to 1cfec16e97f045b446eeb2a86ba8b23e08e792f3
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8258/head:pull/8258` \
`$ git checkout pull/8258`

Update a local copy of the PR: \
`$ git checkout pull/8258` \
`$ git pull https://git.openjdk.java.net/jdk pull/8258/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8258`

View PR using the GUI difftool: \
`$ git pr show -t 8258`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8258.diff">https://git.openjdk.java.net/jdk/pull/8258.diff</a>

</details>

### CODE REVIEW

**Review by theRealAph — APPROVED** [review]
Looks like a good cleanup. VM_Version is getting to be very unruly.

File: src/hotspot/cpu/aarch64/vm_version_aarch64.hpp [hunk 1]
```diff
@@ -100,32 +100,39 @@ class VM_Version : public Abstract_VM_Version {
...
 
+  enum Feature_Flag {
 #define DECLARE_CPU_FEATURE_FLAG(id, name, bit) CPU_##id = (1 << bit),
     CPU_FEATURE_FLAGS(DECLARE_CPU_FEATURE_FLAG)
 #undef DECLARE_CPU_FEATURE_FLAG
   };
 
+  // Feature identification
+#define CPU_FEATURE_DETECTION(id, name, bit) \
+  static bool supports_##name() { return (_features & CPU_##id) != 0; };
```
[the remark this answers is not in the document]
Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.hpp:132]: Yeah, I was thinking this before. Indeed, A53MAC and  STXR_PREFETCH are not CPU feature. Considering that some codes depend on that, it's acceptable to me leaving them here at this moment.

[the remark this answers is not in the document]
Comment (nsjian) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.hpp:132]: OK. Then could you also update the usages of these two `features` with your new functions?

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.hpp:132]: `STXR_PREFETCH` is usually done unconditionally in non-JVM code. Does it ever hurt performance? If not, let's get rid of `STXR_PREFETCH` and do prefetching unconditionally.

[the remark this answers is not in the document]
Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.hpp:132]: @theRealAph 

TBH I don't know the history about this code. Considering that may impact the performance, I don't have too much confidence to dispose of  it in this patch.

Do you think it's fine to trace it in a separate JBS?

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.hpp:132]: Absolutely, yes, getting rid of `STXR_PREFETCH` should be separate from this patch.

We need, as a group, to keep on top of accreting complexity. That's hard when the people changing code don't know the history of that code.

[the remark this answers is not in the document]
Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.hpp:132]: I created a separate JBS for it. https://bugs.openjdk.java.net/browse/JDK-8284990

[the remark this answers is not in the document]
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.hpp:132]: Thanks.

**Review by nsjian — APPROVED** [review]

**Comment by theRealELiu** [comment]
I guess the new name `sha2` caused this issue. My intention was to align the name and CPU feature. I checked the related code but missed this test case at that time. Do you have plan to fix it? Or I can withdraw the name back certainly.

**Pull request closed without merging**